apps/src/componentLibrary/common/helpers.ts

```diff
@@ -21,3 +23,71 @@ export const getAriaPropsFromProps = (props: {
 
   return ariaProps as AriaAttributes;
 };
+
+export const updatePositionedElementStyles = ({
+  nodePosition,
+  positionedElementRef,
+  direction,
+  setPositionedElementStyles,
+  tailOffset,
+  tailLength,
+  isPositionFixed = false,
+}: {
+  nodePosition: HTMLElement | null;
+  positionedElementRef: React.RefObject<HTMLDivElement | null>;
+  direction?: ComponentPlacementDirection | 'none';
+  setPositionedElementStyles: React.Dispatch<
+    React.SetStateAction<React.CSSProperties>
+  >;
+  tailOffset: number;
+  tailLength: number;
+  isPositionFixed?: boolean;
+}) => {
+  if (nodePosition && positionedElementRef.current && direction !== 'none') {
+    const rect = nodePosition.getBoundingClientRect();
+    const tooltipRect = positionedElementRef.current.getBoundingClientRect();
+    const scrollY = isPositionFixed ? 0 : window.scrollY;
+    const scrollX = isPositionFixed ? 0 : window.scrollX;
+    const textDirection = document.documentElement.dir || 'ltr'; // Default to 'ltr' if not specified
+    const isLtr = textDirection === 'ltr';
+
+    const styles: React.CSSProperties = {};
+
+    const verticalMiddlePosition =
+      rect.top + scrollY + rect.height / 2 - tooltipRect.height / 2;
+    const verticalTopPosition =
+      rect.top + scrollY - tooltipRect.height - tailOffset - tailLength;
+    const verticalBottomPosition =
+      rect.top + rect.height + scrollY + tailOffset + tailLength;
+
+    const horizontalMiddlePosition =
+      rect.left + scrollX + rect.width / 2 - tooltipRect.width / 2;
+    const horizontalLeftPosition =
+      rect.left + scrollX - tooltipRect.width - tailOffset - tailLength;
+    const horizontalRightPosition =
+      rect.right + scrollX + tailOffset + tailLength;
+
+    // Calculate the tooltip position based on the direction and its tail length
+    switch (direction) {
+      case 'onRight':
+        styles.top = verticalMiddlePosition;
+        styles.left = isLtr ? horizontalRightPosition : horizontalLeftPosition;
+        break;
+      case 'onBottom':
+        styles.top = verticalBottomPosition;
+        styles.left = horizontalMiddlePosition;
+        break;
+      case 'onLeft':
+        styles.top = verticalMiddlePosition;
+        styles.left = isLtr ? horizontalLeftPosition : horizontalRightPosition;
+        break;
+      case 'onTop':
+      default:
+        styles.top = verticalTopPosition;
+        styles.left = horizontalMiddlePosition;
+        break;
+    }
+
+    setPositionedElementStyles(styles);
```

Comment:
minor: would it be simpler to have this function just return the computed styles, rather than passing through a callback?

Reply:
Agree it makes sense to return the computed styles, also think it makes sense to leave the shortcut for easier updating the styles in multiple places. Splitted this function into 2 functions: `calculatePositionedElementStyles` returns computed styles, `updatePositionedElementStyles` calls `calculatePositionedElementStyles` and updates the styles in callee component react state